src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/StructMarshallingStubMapNode.cs

```diff
@@ -0,0 +1,108 @@
+    /// </summary>
+    internal sealed class StructMarshallingStubMapNode : ObjectNode, ISymbolDefinitionNode
+    {
+        private ObjectAndOffsetSymbolNode _endSymbol;
+        private ExternalReferencesTableNode _externalReferences;
+
+        public StructMarshallingStubMapNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__struct_marshalling_stub_map_End", true);
+            _externalReferences = externalReferences;
+        }
+
+        public ISymbolDefinitionNode EndSymbol => _endSymbol;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__struct_marshalling_stub_map");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => _externalReferences.Section;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName(NodeFactory factory) => this.GetMangledName(factory.NameMangler);
+
+        public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+        {
+            // This node does not trigger generation of other nodes.
+            if (relocsOnly)
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolDefinitionNode[] { this });
+
+            var writer = new NativeWriter();
+            var typeMapHashTable = new VertexHashtable();
+
+            Section hashTableSection = writer.NewSection();
+            hashTableSection.Place(typeMapHashTable);
+
+            foreach (var structEntry in factory.InteropStubManager.GetStructMarshallingTypes())
+            {
+
+                // the order of data written is as follows:
+                //  0. managed struct type
+                //  1. struct marshalling thunk
+                //  2. struct unmarshalling thunk
+                //  3. struct cleanup thunk
+                //  4. uint = NumFields:11 | Size:20 | HasInvalidLayout:1
+                //  5  for each field
+                //      a. name
+                //      b. offset
+
+                var structType = structEntry.StructType;
+                var nativeType = structEntry.NativeStructType;
+                Vertex thunks= writer.GetTuple(
+                    writer.GetUnsignedConstant(_externalReferences.GetIndex(factory.MethodEntrypoint(structEntry.MarshallingThunk))),
+                    writer.GetUnsignedConstant(_externalReferences.GetIndex(factory.MethodEntrypoint(structEntry.UnmarshallingThunk))),
+                    writer.GetUnsignedConstant(_externalReferences.GetIndex(factory.MethodEntrypoint(structEntry.CleanupThunk)))
+                    );
+
+                uint mask = (uint)((nativeType.Fields.Length & 0xEFF) << 21) 
```

Comment:
Might be nice to at least assert if these don't fit the mask, but I'm wondering: GetUnsignedConstant actually uses compressed integers (so if the value is e.g. less than 127, it only takes up one byte) - we might get more space efficiency by just keeping these as two separate values (it still does make sense to try to squeeze the one bit value into one of them). Compressed integers get inefficient as the value grows.